From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Mon, 29 Apr 2024 14:18:58 -0400
Subject: [PATCH] Fix Pufferfish and Purpur patches


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java b/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
index bbdda3b5263118ec00847e94dde761617db09df2..3a23bebec1f2b0f61f2ada307a37046c29f8dcbe 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfoImpl.java
@@ -33,6 +33,8 @@ public record ServerBuildInfoImpl(
     private static final String BRAND_PAPER_NAME = "Paper";
     private static final String BRAND_GALE_NAME = "Gale"; // Gale - branding changes
     // Leaf start - Rebrand
+    private static final String BRAND_PUFFERFISH_NAME = "Pufferfish";
+    private static final String BRAND_PURPUR_NAME = "Purpur";
     private static final String BRAND_LEAF_NAME = "Leaf";
     // Leaf end- Rebrand
 
@@ -68,7 +70,9 @@ public record ServerBuildInfoImpl(
         return brandId.equals(this.brandId)
             || brandId.equals(BRAND_PAPER_ID)
             // Leaf start - Rebrand
-            || brandId.equals(BRAND_GALE_ID); // Gale - branding changes
+            || brandId.equals(BRAND_GALE_ID) // Gale - branding changes
+            || brandId.equals(BRAND_PUFFERFISH_ID)
+            || brandId.equals(BRAND_PURPUR_ID);
             // Leaf end - Rebrand
     }
 
