From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Thu, 20 Mar 2025 00:16:49 +0100
Subject: [PATCH] Fix SparklyPaper - skip EntityScheduler's executeTick checks
 if there isn't any tasks to be run


diff --git a/src/main/java/io/papermc/paper/threadedregions/EntityScheduler.java b/src/main/java/io/papermc/paper/threadedregions/EntityScheduler.java
index bb56c56cdbd8a15803e85412b9c15b59a28e9e59..59becabc96366d97abbc3edf388bfaab1b1461da 100644
--- a/src/main/java/io/papermc/paper/threadedregions/EntityScheduler.java
+++ b/src/main/java/io/papermc/paper/threadedregions/EntityScheduler.java
@@ -190,4 +190,9 @@ public final class EntityScheduler {
             }
         }
     }
+    public boolean hasTasks() {
+        synchronized (this.stateLock) {
+            return !this.currentlyExecuting.isEmpty() || !this.oneTimeDelayed.isEmpty();
+        }
+    }
 }
