From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Thu, 11 Aug 2022 04:13:27 +0800
Subject: [PATCH] Leaf config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 7dbfff8e0d86ba4533568a8f4182506d8a065b63..f0f5bf5d75d58bdd55cff34fca48945380545d20 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2369,6 +2369,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Paper end
 
+        // Leaf start - Leaf config - API
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getLeafConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Leaf end - Leaf config - API
+
         /**
          * Sends the component to the player
          *
