From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Sun, 7 Jul 2024 01:59:11 +0800
Subject: [PATCH] Fix-MC-223153

Related MC issue: https://bugs.mojang.com/browse/MC-223153

diff --git a/net/minecraft/world/level/block/Blocks.java b/net/minecraft/world/level/block/Blocks.java
index bf047be5b577b0d1bf70458df14618bcfe2d1de2..07a8fbfa7eb6e684ea699f009ce2d19311994e39 100644
--- a/net/minecraft/world/level/block/Blocks.java
+++ b/net/minecraft/world/level/block/Blocks.java
@@ -6632,6 +6632,7 @@ public class Blocks {
             .mapColor(MapColor.COLOR_ORANGE)
             .instrument(NoteBlockInstrument.BASEDRUM)
             .requiresCorrectToolForDrops()
+            .sound(SoundType.COPPER) // Leaf - Fix MC-223153
             .strength(5.0F, 6.0F)
     );
     public static final Block RAW_GOLD_BLOCK = register(
