From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Thu, 17 Oct 2024 01:51:38 -0400
Subject: [PATCH] Remove stream and double iteration in enough deep sleeping
 player check


diff --git a/net/minecraft/server/players/SleepStatus.java b/net/minecraft/server/players/SleepStatus.java
index 2a7ae521654ad5c9f392baa5562e64bb71b13097..2d0033b18f46fb05df12536d68b4bce455abfeed 100644
--- a/net/minecraft/server/players/SleepStatus.java
+++ b/net/minecraft/server/players/SleepStatus.java
@@ -15,9 +15,24 @@ public class SleepStatus {
 
     public boolean areEnoughDeepSleeping(int requiredSleepPercentage, List<ServerPlayer> sleepingPlayers) {
         // CraftBukkit start
-        int i = (int) sleepingPlayers.stream().filter(player -> player.isSleepingLongEnough() || player.fauxSleeping).count();
-        boolean anyDeepSleep = sleepingPlayers.stream().anyMatch(Player::isSleepingLongEnough);
-        return anyDeepSleep && i >= this.sleepersNeeded(requiredSleepPercentage);
+        // Leaf start - Remove stream and double iteration in enough deep sleeping player check
+        int count = 0;
+        boolean anyPlayerSleeping = false;
+
+        for (ServerPlayer player : sleepingPlayers) {
+            final boolean isSleepingLongEnough = player.isSleepingLongEnough();
+
+            if (isSleepingLongEnough) {
+                anyPlayerSleeping = true;
+            }
+
+            if (isSleepingLongEnough || player.fauxSleeping || (player.level().purpurConfig.idleTimeoutCountAsSleeping && player.isAfk())) { // Purpur - AFK API
+                count++;
+            }
+        }
+
+        return anyPlayerSleeping && count >= this.sleepersNeeded(requiredSleepPercentage);
+        // Leaf end - Remove stream and double iteration in enough deep sleeping player check
         // CraftBukkit end
     }
 
