From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Sat, 22 Mar 2025 21:23:45 +0100
Subject: [PATCH] fix NPE on MultifaceSpreader


diff --git a/net/minecraft/world/level/block/MultifaceSpreader.java b/net/minecraft/world/level/block/MultifaceSpreader.java
index 60f47334bb855d5216f57f888f131ba41f728d21..93ab8bbc7c76428d7fa3bb38567dd6f3f007d5f2 100644
--- a/net/minecraft/world/level/block/MultifaceSpreader.java
+++ b/net/minecraft/world/level/block/MultifaceSpreader.java
@@ -10,6 +10,8 @@ import net.minecraft.world.level.BlockGetter;
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.block.state.BlockState;
 
+import static org.dreeam.leaf.config.LeafConfig.LOGGER;
+
 public class MultifaceSpreader {
     public static final MultifaceSpreader.SpreadType[] DEFAULT_SPREAD_ORDER = new MultifaceSpreader.SpreadType[]{
         MultifaceSpreader.SpreadType.SAME_POSITION, MultifaceSpreader.SpreadType.SAME_PLANE, MultifaceSpreader.SpreadType.WRAP_AROUND
@@ -148,13 +150,19 @@ public class MultifaceSpreader {
         }
 
         default boolean placeBlock(LevelAccessor level, MultifaceSpreader.SpreadPos pos, BlockState state, boolean markForPostprocessing) {
+            // Check for null
+            if (pos.source() == null || pos.pos() == null) {
+                LOGGER.warn("Invalid SpreadPos with null source or position: {}", pos);
+                return false;
+            }
+
             BlockState stateForPlacement = this.getStateForPlacement(state, level, pos.pos(), pos.face());
             if (stateForPlacement != null) {
                 if (markForPostprocessing) {
                     level.getChunk(pos.pos()).markPosForPostprocessing(pos.pos());
                 }
 
-                return org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, pos.source(), pos.pos(), stateForPlacement, 2); // CraftBukkit
+                return org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, pos.source(), pos.pos(), stateForPlacement, 2);
             } else {
                 return false;
             }
