From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Thu, 8 May 2025 04:56:30 +0900
Subject: [PATCH] send and flush position while knockback player


diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 477455fdfcc591a89823e88983eb12dabb078d9b..fdc9c68a0ed30b8b342ceaca5d988e82524d6bf9 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -1411,6 +1411,20 @@ public abstract class Player extends LivingEntity {
                                     player.setVelocity(event.getVelocity());
                                 }
 
+
+                                // Leaf start
+                                if (org.dreeam.leaf.config.modules.gameplay.Knockback.flushKnockback && !cancelled && this instanceof ServerPlayer player1) {
+                                    ServerPlayer target1 = (ServerPlayer) target;
+                                    var targetV = target1.getDeltaMovement();
+                                    target1.aiStep();
+                                    target1.setDeltaMovement(targetV);
+                                    target1.connection.send(net.minecraft.network.protocol.game.ClientboundEntityPositionSyncPacket.of(this));
+                                    player1.connection.send(net.minecraft.network.protocol.game.ClientboundEntityPositionSyncPacket.of(target1));
+                                    player1.connection.send(new ClientboundSetEntityMotionPacket(target1));
+                                    player1.hasImpulse = true;
+                                    target1.hasImpulse = true;
+                                }
+                                // Leaf end
                                 if (!cancelled) {
                                 ((ServerPlayer)target).connection.send(new ClientboundSetEntityMotionPacket(target));
                                 target.hurtMarked = false;
@@ -1480,6 +1494,12 @@ public abstract class Player extends LivingEntity {
                             }
 
                             this.causeFoodExhaustion(this.level().spigotConfig.combatExhaustion, org.bukkit.event.entity.EntityExhaustionEvent.ExhaustionReason.ATTACK); // CraftBukkit - EntityExhaustionEvent // Spigot - Change to use configurable value
+                            // Leaf start
+                            if (org.dreeam.leaf.config.modules.gameplay.Knockback.flushKnockback && target instanceof ServerPlayer target1 && this instanceof ServerPlayer player1) {
+                                target1.connection.connection.flushChannel();
+                                player1.connection.connection.flushChannel();
+                            }
+                            // Leaf end
                         } else {
                             this.sendSoundEffect(this, this.getX(), this.getY(), this.getZ(), SoundEvents.PLAYER_ATTACK_NODAMAGE, this.getSoundSource(), 1.0F, 1.0F); // Paper - send while respecting visibility
                             // CraftBukkit start - resync on cancelled event
