From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Thu, 8 May 2025 04:56:30 +0900
Subject: [PATCH] serverside move and flush while knockback player


diff --git a/net/minecraft/world/entity/player/Player.java b/net/minecraft/world/entity/player/Player.java
index 477455fdfcc591a89823e88983eb12dabb078d9b..41fbd978632572636ca25a279fd15f818fa4eb22 100644
--- a/net/minecraft/world/entity/player/Player.java
+++ b/net/minecraft/world/entity/player/Player.java
@@ -1414,6 +1414,22 @@ public abstract class Player extends LivingEntity {
                                 if (!cancelled) {
                                 ((ServerPlayer)target).connection.send(new ClientboundSetEntityMotionPacket(target));
                                 target.hurtMarked = false;
+                                    // Leaf start
+                                    if (org.dreeam.leaf.config.modules.gameplay.Knockback.flushKnockback) {
+                                        ServerPlayer targetPlayer = (ServerPlayer) target;
+                                        targetPlayer.travel(targetPlayer.getDeltaMovement());
+                                        targetPlayer.connection.send(new net.minecraft.network.protocol.game.ClientboundSetEntityMotionPacket(this));
+                                        targetPlayer.connection.send(net.minecraft.network.protocol.game.ClientboundEntityPositionSyncPacket.of(this));
+                                        targetPlayer.connection.connection.flushChannel();
+                                        targetPlayer.hasImpulse = true;
+                                        if (this instanceof ServerPlayer player1) {
+                                            player1.connection.send(new net.minecraft.network.protocol.game.ClientboundSetEntityMotionPacket(targetPlayer));
+                                            player1.connection.send(net.minecraft.network.protocol.game.ClientboundEntityPositionSyncPacket.of(targetPlayer));
+                                            player1.connection.connection.flushChannel();
+                                            player1.hasImpulse = true;
+                                        }
+                                    }
+                                    // Leaf end
                                 target.setDeltaMovement(deltaMovement);
                                 }
                                 // CraftBukkit end
