From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Thu, 13 Feb 2025 01:25:40 +0100
Subject: [PATCH] Remove iterators from Inventory


diff --git a/net/minecraft/world/entity/player/Inventory.java b/net/minecraft/world/entity/player/Inventory.java
index d9cb4f0ed0c4f63362c837aeef3c4194911455c9..a36d5f486e4b936e9107aa0dce263ad8afc30373 100644
--- a/net/minecraft/world/entity/player/Inventory.java
+++ b/net/minecraft/world/entity/player/Inventory.java
@@ -439,13 +439,15 @@ public class Inventory implements Container, Nameable {
             }
         }
 
-        for (EquipmentSlot equipmentSlot : EQUIPMENT_SLOT_MAPPING.values()) {
+        // Leaf start - Remove iterators from Inventory
+        for (EquipmentSlot equipmentSlot : EQUIPMENT_SLOTS_SORTED_BY_INDEX) {
             ItemStack itemStack = this.equipment.get(equipmentSlot);
             if (itemStack == stack) {
                 this.equipment.set(equipmentSlot, ItemStack.EMPTY);
                 return;
             }
         }
+        // Leaf end - Remove iterators from Inventory
     }
 
     @Override
@@ -504,17 +506,20 @@ public class Inventory implements Container, Nameable {
 
     @Override
     public boolean isEmpty() {
-        for (ItemStack itemStack : this.items) {
+        // Leaf start - Remove iterators from Inventory
+        for (int i = 0; i < this.items.size(); i++) {
+            ItemStack itemStack = this.items.get(i);
             if (!itemStack.isEmpty()) {
                 return false;
             }
         }
 
-        for (EquipmentSlot equipmentSlot : EQUIPMENT_SLOT_MAPPING.values()) {
+        for (EquipmentSlot equipmentSlot : EQUIPMENT_SLOTS_SORTED_BY_INDEX) {
             if (!this.equipment.get(equipmentSlot).isEmpty()) {
                 return false;
             }
         }
+        // Leaf end - Remove iterators from Inventory
 
         return true;
     }
@@ -561,31 +566,58 @@ public class Inventory implements Container, Nameable {
     }
 
     public boolean contains(ItemStack stack) {
-        for (ItemStack itemStack : this) {
+        // Leaf start - Remove iterators from Inventory
+        for (int i = 0; i < this.items.size(); i++) {
+            ItemStack itemStack = this.items.get(i);
             if (!itemStack.isEmpty() && ItemStack.isSameItemSameComponents(itemStack, stack)) {
                 return true;
             }
         }
+        for (EquipmentSlot equipmentSlot : EQUIPMENT_SLOTS_SORTED_BY_INDEX) {
+            ItemStack itemStack = this.equipment.get(equipmentSlot);
+            if (!itemStack.isEmpty() && ItemStack.isSameItemSameComponents(itemStack, stack)) {
+                return true;
+            }
+        }
+        // Leaf end - Remove iterators from Inventory
 
         return false;
     }
 
     public boolean contains(TagKey<Item> tag) {
-        for (ItemStack itemStack : this) {
+        // Leaf start - Remove iterators from Inventory
+        for (int i = 0; i < this.items.size(); i++) {
+            ItemStack itemStack = this.items.get(i);
+            if (!itemStack.isEmpty() && itemStack.is(tag)) {
+                return true;
+            }
+        }
+        for (EquipmentSlot equipmentSlot : EQUIPMENT_SLOTS_SORTED_BY_INDEX) {
+            ItemStack itemStack = this.equipment.get(equipmentSlot);
             if (!itemStack.isEmpty() && itemStack.is(tag)) {
                 return true;
             }
         }
+        // Leaf end - Remove iterators from Inventory
 
         return false;
     }
 
     public boolean contains(Predicate<ItemStack> predicate) {
-        for (ItemStack itemStack : this) {
+        // Leaf start - Remove iterators from Inventory
+        for (int i = 0; i < this.items.size(); i++) {
+            ItemStack itemStack = this.items.get(i);
             if (predicate.test(itemStack)) {
                 return true;
             }
         }
+        for (EquipmentSlot equipmentSlot : EQUIPMENT_SLOTS_SORTED_BY_INDEX) {
+            ItemStack itemStack = this.equipment.get(equipmentSlot);
+            if (predicate.test(itemStack)) {
+                return true;
+            }
+        }
+        // Leaf end - Remove iterators from Inventory
 
         return false;
     }
@@ -605,9 +637,12 @@ public class Inventory implements Container, Nameable {
     }
 
     public void fillStackedContents(StackedItemContents contents) {
-        for (ItemStack itemStack : this.items) {
+        // Leaf start - Remove iterators from Inventory
+        for (int i = 0; i < this.items.size(); i++) {
+            ItemStack itemStack = this.items.get(i);
             contents.accountSimpleStack(itemStack);
         }
+        // Leaf end - Remove iterators from Inventory
     }
 
     public ItemStack removeFromSelected(boolean removeStack) {
