From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Sat, 17 May 2025 19:01:50 +0900
Subject: [PATCH] Optimize getScaledTrackingDistance


diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index 574765713c194a6cf3eb7c125e78ed77c82aba2e..980d18434a79d1ff5e16b71aeaf34b486e665f9d 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -786,7 +786,13 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
 
     @Override
     public int getScaledTrackingDistance(int trackingDistance) {
-        return this.getProperties().entityBroadcastRangePercentage * trackingDistance / 100;
+        // Leaf start - Optimize getScaledTrackingDistance
+        int p = this.getProperties().entityBroadcastRangePercentage;
+        if (p == 100) {
+            return trackingDistance;
+        }
+        return p * trackingDistance / 100;
+        // Leaf end - Optimize getScaledTrackingDistance
     }
 
     @Override
