From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Fri, 28 Oct 2022 17:48:45 -0400
Subject: [PATCH] Bump Dependencies

TODO - Dreeam: Bump & test dependencies, impl new features?

diff --git a/build.gradle.kts b/build.gradle.kts
index 3b0cd45cb07d9563c84901729f1f7edc498653bd..c35ffb8565218d9368cc583ae96ea86424184eed 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -44,28 +44,32 @@ dependencies {
           all its classes to check if they are plugins.
           Scanning takes about 1-2 seconds so adding this speeds up the server start.
      */
-    implementation("org.apache.logging.log4j:log4j-core:2.19.0") // Paper - implementation
-    log4jPlugins.annotationProcessorConfigurationName("org.apache.logging.log4j:log4j-core:2.19.0") // Paper - Needed to generate meta for our Log4j plugins
+    // Leaf start - Bump Dependencies
+    implementation("org.apache.logging.log4j:log4j-core:2.24.2") // Paper - implementation
+    log4jPlugins.annotationProcessorConfigurationName("org.apache.logging.log4j:log4j-core:2.24.2") // Paper - Needed to generate meta for our Log4j plugins
     runtimeOnly(log4jPlugins.output)
     alsoShade(log4jPlugins.output)
-    implementation("io.netty:netty-codec-haproxy:4.1.97.Final") // Paper - Add support for proxy protocol
+    implementation("io.netty:netty-codec-haproxy:4.1.115.Final") // Paper - Add support for proxy protocol
     // Paper end
-    implementation("org.apache.logging.log4j:log4j-iostreams:2.22.1") // Paper - remove exclusion
+    implementation("org.apache.logging.log4j:log4j-iostreams:2.24.2") // Paper - remove exclusion
+    // Leaf end - Bump Dependencies
     implementation("org.ow2.asm:asm-commons:9.7.1")
     implementation("org.spongepowered:configurate-yaml:4.2.0-SNAPSHOT") // Paper - config files
     implementation("commons-lang:commons-lang:2.6")
-    runtimeOnly("org.xerial:sqlite-jdbc:3.46.1.3")
+    runtimeOnly("org.xerial:sqlite-jdbc:3.47.1.0") // Leaf - Bump Dependencies
     runtimeOnly("com.mysql:mysql-connector-j:9.1.0")
-    runtimeOnly("com.lmax:disruptor:3.4.4") // Paper
+    runtimeOnly("com.lmax:disruptor:3.4.4") // Paper // Dreeam TODO - Waiting Log4j 3.x to support disruptor 4.0.0
     // Paper start - Use Velocity cipher
-    implementation("com.velocitypowered:velocity-native:3.3.0-SNAPSHOT") {
+    implementation("com.velocitypowered:velocity-native:3.4.0-SNAPSHOT") { // Leaf - Bump Dependencies
         isTransitive = false
     }
     // Paper end - Use Velocity cipher
 
-    runtimeOnly("org.apache.maven:maven-resolver-provider:3.9.6")
-    runtimeOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.9.18")
-    runtimeOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.9.18")
+    // Leaf start - Bump Dependencies
+    runtimeOnly("org.apache.maven:maven-resolver-provider:3.9.9")
+    runtimeOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.9.22") // Dreeam TODO - Update to 2.0.1
+    runtimeOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.9.22") // Dreeam TODO - Update to 2.0.1
+    // Leaf end
 
     // Purpur start
     implementation("org.mozilla:rhino-runtime:1.7.15")
@@ -73,14 +77,16 @@ dependencies {
     implementation("dev.omega24:upnp4j:1.0")
     // Purpur end
 
-    testImplementation("io.github.classgraph:classgraph:4.8.47") // Paper - mob goal test
-    testImplementation("org.junit.jupiter:junit-jupiter:5.10.2")
-    testImplementation("org.junit.platform:junit-platform-suite-engine:1.10.0")
-    testImplementation("org.hamcrest:hamcrest:2.2")
-    testImplementation("org.mockito:mockito-core:5.14.1")
-    mockitoAgent("org.mockito:mockito-core:5.14.1") { isTransitive = false } // Paper - configure mockito agent that is needed in newer java versions
+    // Leaf start - Bump Dependencies
+    testImplementation("io.github.classgraph:classgraph:4.8.179") // Paper - mob goal test
+    testImplementation("org.junit.jupiter:junit-jupiter:5.11.3")
+    testImplementation("org.junit.platform:junit-platform-suite-engine:1.11.3")
+    testImplementation("org.hamcrest:hamcrest:3.0")
+    testImplementation("org.mockito:mockito-core:5.14.2")
+    mockitoAgent("org.mockito:mockito-core:5.14.2") { isTransitive = false } // Paper - configure mockito agent that is needed in newer java versions
     testImplementation("org.ow2.asm:asm-tree:9.7.1")
-    testImplementation("org.junit-pioneer:junit-pioneer:2.2.0") // Paper - CartesianTest
+    testImplementation("org.junit-pioneer:junit-pioneer:2.3.0") // Paper - CartesianTest
+    // Leaf end - Bump Dependencies
     implementation("net.neoforged:srgutils:1.0.9") // Paper - mappings handling
     implementation("net.neoforged:AutoRenamingTool:2.0.3") // Paper - remap plugins
     // Paper start - Remap reflection
@@ -93,6 +99,8 @@ dependencies {
     implementation("me.lucko:spark-api:0.1-20240720.200737-2")
     implementation("me.lucko:spark-paper:1.10.119-SNAPSHOT")
     // Paper end - spark
+
+    implementation("io.netty:netty-all:4.1.115.Final") // Leaf - Bump Dependencies // Dreeam TODO - Update to 4.2.0
 }
 
 paperweight {
@@ -295,3 +303,8 @@ sourceSets {
     }
 }
 // Gale end - package license into jar
+// Leaf start - Bump Dependencies
+repositories {
+    mavenCentral()
+}
+// Leaf end - Bump Dependencies
