From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: lilingfengdev <145678359+lilingfengdev@users.noreply.github.com>
Date: Thu, 18 Jan 2024 13:30:02 +0800
Subject: [PATCH] Polpot: Make egg and snowball can knockback player

Original project: https://github.com/HaHaWTH/Polpot

diff --git a/src/main/java/net/minecraft/world/entity/projectile/Snowball.java b/src/main/java/net/minecraft/world/entity/projectile/Snowball.java
index 0db58e7d63a5c1b43a2224c247979f23a1d3f899..c2ffa86e0492ed1aa6a1850e8da49fdca2c2e2c7 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/Snowball.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/Snowball.java
@@ -61,6 +61,12 @@ public class Snowball extends ThrowableItemProjectile {
         int i = entity.level().purpurConfig.snowballDamage >= 0 ? entity.level().purpurConfig.snowballDamage : entity instanceof Blaze ? 3 : 0; // Purpur
 
         entity.hurt(this.damageSources().thrown(this, this.getOwner()), (float) i);
+        // Leaf start - Polpot - Make snowball can knockback player
+        if (org.dreeam.leaf.config.modules.gameplay.Knockback.snowballCanKnockback && entity instanceof net.minecraft.server.level.ServerPlayer) {
+            entity.hurt(this.damageSources().thrown(this, this.getOwner()), 0.0000001F);
+            ((net.minecraft.server.level.ServerPlayer) entity).knockback(0.4000000059604645D, this.getX() - entity.getX(), this.getZ() - entity.getZ());
+        }
+        // Leaf end - Polpot - Make snowball can knockback player
     }
 
     // Purpur start - borrowed and modified code from ThrownPotion#onHitBlock and ThrownPotion#dowseFire
diff --git a/src/main/java/net/minecraft/world/entity/projectile/ThrownEgg.java b/src/main/java/net/minecraft/world/entity/projectile/ThrownEgg.java
index 155c2bbd35adacb7c3668fbe81a7c454e5102c8b..15de9bf1c39c9cc44debd196725e1f7ebf247e60 100644
--- a/src/main/java/net/minecraft/world/entity/projectile/ThrownEgg.java
+++ b/src/main/java/net/minecraft/world/entity/projectile/ThrownEgg.java
@@ -51,7 +51,14 @@ public class ThrownEgg extends ThrowableItemProjectile {
     @Override
     protected void onHitEntity(EntityHitResult entityHitResult) {
         super.onHitEntity(entityHitResult);
+        Entity entity = entityHitResult.getEntity(); // Polpot - make egg can knockback player
         entityHitResult.getEntity().hurt(this.damageSources().thrown(this, this.getOwner()), 0.0F);
+        // Leaf start - Polpot - Make egg can knockback player
+        if (org.dreeam.leaf.config.modules.gameplay.Knockback.eggCanKnockback && entity instanceof ServerPlayer) {
+            entity.hurt(this.damageSources().thrown(this, this.getOwner()), 0.0000001F);
+            ((ServerPlayer) entity).knockback(0.4000000059604645D, this.getX() - entity.getX(), this.getZ() - entity.getZ());
+        }
+        // Leaf end - Polpot - Make egg can knockback player
     }
 
     @Override
diff --git a/src/main/java/org/dreeam/leaf/config/modules/gameplay/Knockback.java b/src/main/java/org/dreeam/leaf/config/modules/gameplay/Knockback.java
new file mode 100644
index 0000000000000000000000000000000000000000..7b434158818834450ea43611e2ab5636917bb938
--- /dev/null
+++ b/src/main/java/org/dreeam/leaf/config/modules/gameplay/Knockback.java
@@ -0,0 +1,28 @@
+package org.dreeam.leaf.config.modules.gameplay;
+
+import org.dreeam.leaf.config.ConfigModules;
+import org.dreeam.leaf.config.EnumConfigCategory;
+
+public class Knockback extends ConfigModules {
+
+    public String getBasePath() {
+        return EnumConfigCategory.GAMEPLAY.getBaseKeyName() + ".knockback";
+    }
+
+    public static boolean snowballCanKnockback = false;
+    public static boolean eggCanKnockback = false;
+
+    @Override
+    public void onLoaded() {
+        snowballCanKnockback = config.getBoolean(getBasePath() + ".snowball-knockback-players", snowballCanKnockback,
+                config.pickStringRegionBased(
+                        "Make snowball can knockback players.",
+                        "使雪球可以击退玩家."
+                        ));
+        eggCanKnockback = config.getBoolean(getBasePath() + ".egg-knockback-players", eggCanKnockback,
+                config.pickStringRegionBased(
+                        "Make egg can knockback players.",
+                        "使鸡蛋可以击退玩家."
+                        ));
+    }
+}
