From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Tue, 27 Feb 2024 03:17:10 -0500
Subject: [PATCH] Remove useless creating stats json bases on player name logic


diff --git a/src/main/java/net/minecraft/server/players/PlayerList.java b/src/main/java/net/minecraft/server/players/PlayerList.java
index 0d41cd4fe9b423a8644475494bcfb73f98e8b70b..8d8f552acd4d2e846e1fd0bc19574d93c372fc2b 100644
--- a/src/main/java/net/minecraft/server/players/PlayerList.java
+++ b/src/main/java/net/minecraft/server/players/PlayerList.java
@@ -1687,6 +1687,8 @@ public abstract class PlayerList {
             File file = this.server.getWorldPath(LevelResource.PLAYER_STATS_DIR).toFile();
             File file1 = new File(file, String.valueOf(uuid) + ".json");
 
+            // Leaf start - Remove useless creating stats json bases on player name logic
+            /*
             if (!file1.exists()) {
                 File file2 = new File(file, displayName + ".json"); // CraftBukkit
                 Path path = file2.toPath();
@@ -1695,6 +1697,8 @@ public abstract class PlayerList {
                     file2.renameTo(file1);
                 }
             }
+             */
+            // Leaf end - Remove useless creating stats json bases on player name logic
 
             serverstatisticmanager = new ServerStatsCounter(this.server, file1);
             // this.stats.put(uuid, serverstatisticmanager); // CraftBukkit
