From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Tue, 17 Sep 2024 02:26:44 -0400
Subject: [PATCH] Remove stream in entity visible effects filter


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index d0755333e86009bbb584481f4796cfc60643a53f..92b253e0f96c742e02a8c8316cd669b949cabf42 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -1008,7 +1008,15 @@ public abstract class LivingEntity extends Entity implements Attackable {
     }
 
     private void updateSynchronizedMobEffectParticles() {
-        List<ParticleOptions> list = this.activeEffects.values().stream().filter(MobEffectInstance::isVisible).map(MobEffectInstance::getParticleOptions).toList();
+        // Leaf start - Remove stream in entity visible effects filter
+        List<ParticleOptions> list = new ArrayList<>();
+
+        for (MobEffectInstance effect : this.activeEffects.values()) {
+            if (effect.isVisible()) {
+                list.add(effect.getParticleOptions());
+            }
+        }
+        // Leaf end - Remove stream in entity visible effects filter
 
         this.entityData.set(LivingEntity.DATA_EFFECT_PARTICLES, list);
         this.entityData.set(LivingEntity.DATA_EFFECT_AMBIENCE_ID, LivingEntity.areAllEffectsAmbient(this.activeEffects.values()));
