From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Tue, 17 Sep 2024 02:26:44 -0400
Subject: [PATCH] Remove stream in entity visible effects filter


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 291ae21d002a4215407d18806c94a51e9c32a7c2..9e771e3aa2dbf5afb1f5fe9c3d9e6f16031f8df1 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -1010,7 +1010,15 @@ public abstract class LivingEntity extends Entity implements Attackable {
     }
 
     private void updateSynchronizedMobEffectParticles() {
-        List<ParticleOptions> list = this.activeEffects.values().stream().filter(MobEffectInstance::isVisible).map(MobEffectInstance::getParticleOptions).toList();
+        // Leaf start - Remove stream in entity visible effects filter
+        List<ParticleOptions> list = new ArrayList<>();
+
+        for (MobEffectInstance effect : this.activeEffects.values()) {
+            if (effect.isVisible()) {
+                list.add(effect.getParticleOptions());
+            }
+        }
+        // Leaf end - Remove stream in entity visible effects filter
 
         this.entityData.set(LivingEntity.DATA_EFFECT_PARTICLES, list);
         this.entityData.set(LivingEntity.DATA_EFFECT_AMBIENCE_ID, LivingEntity.areAllEffectsAmbient(this.activeEffects.values()));
