From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: HaHaWTH <102713261+HaHaWTH@users.noreply.github.com>
Date: Sun, 7 Jul 2024 01:59:11 +0800
Subject: [PATCH] Fix-MC-223153

Related MC issue: https://bugs.mojang.com/browse/MC-223153

diff --git a/src/main/java/net/minecraft/world/level/block/Blocks.java b/src/main/java/net/minecraft/world/level/block/Blocks.java
index 63d67d46d30ed8ed57cdc0e59b6cb6b75ab22c1f..539b5625ba0ef7389ff1e7041af86f538640f3d9 100644
--- a/src/main/java/net/minecraft/world/level/block/Blocks.java
+++ b/src/main/java/net/minecraft/world/level/block/Blocks.java
@@ -6611,6 +6611,7 @@ public class Blocks {
             .mapColor(MapColor.COLOR_ORANGE)
             .instrument(NoteBlockInstrument.BASEDRUM)
             .requiresCorrectToolForDrops()
+            .sound(SoundType.COPPER) // Leaf - Fix MC-223153
             .strength(5.0F, 6.0F)
     );
     public static final Block RAW_GOLD_BLOCK = register(
