From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Tue, 17 Sep 2024 02:35:13 -0400
Subject: [PATCH] Replace Entity active effects map with optimized collection

Dreeam TODO: check this

diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 9e771e3aa2dbf5afb1f5fe9c3d9e6f16031f8df1..6d9b94e08fc8cdc5b2f31753484359a4f673a722 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -200,6 +200,10 @@ public abstract class LivingEntity extends Entity implements Attackable {
     public static final String ATTRIBUTES_FIELD = "attributes";
     private final AttributeMap attributes;
     public CombatTracker combatTracker = new CombatTracker(this);
+    // Need to figure out the difference of mem access pattern between hash map and obj2obj hash map (separate chaining vs open addressing)
+    // Benchmark is needed for get calls for this active effects map.
+    // Also need to check whether call from out of main using bukkit api
+    //public final Map<Holder<MobEffect>, MobEffectInstance> activeEffects = new it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap<>(0); // Leaf - Replace Entity active effects map with optimized collection
     public final Map<Holder<MobEffect>, MobEffectInstance> activeEffects = Maps.newHashMap();
     private final NonNullList<ItemStack> lastHandItemStacks;
     private final NonNullList<ItemStack> lastArmorItemStacks;
@@ -1012,8 +1016,9 @@ public abstract class LivingEntity extends Entity implements Attackable {
     private void updateSynchronizedMobEffectParticles() {
         // Leaf start - Remove stream in entity visible effects filter
         List<ParticleOptions> list = new ArrayList<>();
+        final Collection<MobEffectInstance> effectsValues = this.activeEffects.values(); // Leaf - Replace Entity active effects map with optimized collection
 
-        for (MobEffectInstance effect : this.activeEffects.values()) {
+        for (MobEffectInstance effect : effectsValues) { // Leaf - Replace Entity active effects map with optimized collection
             if (effect.isVisible()) {
                 list.add(effect.getParticleOptions());
             }
@@ -1021,7 +1026,7 @@ public abstract class LivingEntity extends Entity implements Attackable {
         // Leaf end - Remove stream in entity visible effects filter
 
         this.entityData.set(LivingEntity.DATA_EFFECT_PARTICLES, list);
-        this.entityData.set(LivingEntity.DATA_EFFECT_AMBIENCE_ID, LivingEntity.areAllEffectsAmbient(this.activeEffects.values()));
+        this.entityData.set(LivingEntity.DATA_EFFECT_AMBIENCE_ID, LivingEntity.areAllEffectsAmbient(effectsValues)); // Leaf - Replace Entity active effects map with optimized collection
     }
 
     private void updateGlowingStatus() {
