From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Fri, 8 Nov 2024 03:50:16 +0100
Subject: [PATCH] Fix-tick-function-tag-running-before-load


diff --git a/src/main/java/net/minecraft/server/ServerFunctionManager.java b/src/main/java/net/minecraft/server/ServerFunctionManager.java
index f9f893a286147c9a8e49f78891381227380a2a14..b7a58ac8db0d6969155710bca162826a5ed0b6ce 100644
--- a/src/main/java/net/minecraft/server/ServerFunctionManager.java
+++ b/src/main/java/net/minecraft/server/ServerFunctionManager.java
@@ -47,7 +47,7 @@ public class ServerFunctionManager {
                 this.executeTagFunctions(collection, ServerFunctionManager.LOAD_FUNCTION_TAG);
             }
 
-            this.executeTagFunctions(this.ticking, ServerFunctionManager.TICK_FUNCTION_TAG);
+            //this.executeTagFunctions(this.ticking, ServerFunctionManager.TICK_FUNCTION_TAG);
         }
     }
 
@@ -74,6 +74,7 @@ public class ServerFunctionManager {
         } catch (Exception exception) {
             ServerFunctionManager.LOGGER.warn("Failed to execute function {}", function.id(), exception);
         }
+        this.executeTagFunctions(this.ticking, ServerFunctionManager.TICK_FUNCTION_TAG); // Mirai - fix tick function tag running before load
 
     }
 
