From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Fri, 8 Nov 2024 04:32:35 +0100
Subject: [PATCH] branchless-clamp-logic


diff --git a/src/main/java/net/minecraft/util/Mth.java b/src/main/java/net/minecraft/util/Mth.java
index ac359e7237ff6256045ee47ca287d9315900dc0e..dd79fd5331f706eb440ce96c4061ae434017967e 100644
--- a/src/main/java/net/minecraft/util/Mth.java
+++ b/src/main/java/net/minecraft/util/Mth.java
@@ -87,7 +87,10 @@ public class Mth {
     }
 
     public static int clamp(int value, int min, int max) {
-        return Math.min(Math.max(value, min), max);
+        // First clamp to min - Leaf
+        value = value < min ? min : value;
+        // Then clamp to max - Leaf
+        return value > max ? max : value;
     }
 
     public static long clamp(long value, long min, long max) {
